src/bodywork/git.py

```diff
@@ -44,102 +50,160 @@ def download_project_code_from_repo(
-        known_hosts.touch(0o700, exist_ok=False)
-        known_hosts.write_text(ssh_public_key)
-
-    os.environ['GIT_SSH_COMMAND'] = (
-        f'ssh -i {private_key} -o UserKnownHostsFile={known_hosts}'
+        try:
+            ssh_dir.mkdir(mode=0o700, exist_ok=True)
+            private_key.touch(0o700, exist_ok=False)
+            private_key.write_text(os.environ[SSH_PRIVATE_KEY_ENV_VAR])
+        except OSError as e:
+            raise RuntimeError(
+                f"Unable to create private key {private_key} from"
+                f" {SSH_PRIVATE_KEY_ENV_VAR} environment variable"
+            ) from e
+
+    try:
+        known_hosts = ssh_dir / "known_hosts"
+        if not known_hosts.exists():
+            known_hosts.touch(0o700, exist_ok=False)
+            known_hosts.write_text(get_ssh_public_key_from_domain(hostname))
+        elif not known_hosts_contains_domain_key(hostname, known_hosts):
+            known_hosts.touch(0o700)
+            with known_hosts.open(mode="a") as file_handle:
+                file_handle.write(get_ssh_public_key_from_domain(hostname))
+    except OSError as e:
+        raise RuntimeError(
+            f"Error updating known hosts with public key from {hostname}"
+        ) from e
+
+    os.environ["GIT_SSH_COMMAND"] = (
+        f"ssh -i '{private_key}'" f" -o UserKnownHostsFile='{known_hosts}'"
     )
+
+
+def known_hosts_contains_domain_key(hostname: str, known_hosts_filepath: Path) -> bool:
+    """Checks to see if the host is in the list of keys in the known_hosts file.
+
+    :param known_hosts_filepath: path to known_hosts file
+    :param hostname: Host name to check for.
+    :return: bool if the hostname is in the file
+    """
+    return hostname in known_hosts_filepath.read_text()
+
+
+def get_ssh_public_key_from_domain(hostname: str) -> str:
+    """Gets the public SSH key from the host and checks if the fingerprint matches
+     the stored fingerprint of valid servers
+
+    :param hostname: Name of host to retrieve the key from e.g. Gitlab.com
+    :return: The public SSH Key of the host.
+    """
+    fingerprints = {
+        "github.com": GITHUB_SSH_FINGERPRINT,
+        "gitlab.com": GITLAB_SSH_FINGERPRINT,
+        "bitbucket.org": BITBUCKET_SSH_FINGERPRINT,
+    }
+    if hostname in fingerprints:
+        try:
+            server_key = run(
+                ["ssh-keyscan", "-t", "rsa", hostname],
+                check=True,
+                capture_output=True,
+                encoding="utf-8",
+            ).stdout
+            fingerprint = run(
+                ["ssh-keygen", "-l", "-f", "-"],
```

Comment:
What's does the `"-"` do?

I couldn't work it out from reading the ssh-keygen man pages.

Reply:
It won't be in there, it's required for piping in the input from the previous command using the input=server_key.

Reply:
Is that a generic Linux thing?

Is there any chance of making it more explicit, because I'll  never remember that.

Reply:
I think it is and it works on Windows as well. I don't really know how to make it more explicit apart from in the comments. Would an explanation in the comment suffice?

Reply:
A comment at the very least please.

I thought you could just pass is the filename from the previous command that it's implicitly referring to. Have I misunderstood what's going on?

Reply:
That would involve creating a file from the previous command and I wanted to avoid creating a file just for the sake of passing it into the keygen command. I think this is the more robust solution as it removes any file handling.